raimitigations/dataprocessing/imputer/knn_imputer.py

```diff
@@ -0,0 +1,262 @@
+                    + "If you'd like to ordinal encode and impute these columns, use 'enable_encoder'=True.\n"
+                    + "Note that encoded columns are not guaranteed to reverse transform if they have imputed values.\n"
+                    + "If you'd like to use a different type of encoding before imputation, consider using the Pipeline class and call your own encoder before calling this subclass."
+                )
+            df_to_transf = df_valid
+
+        else:
+            warnings.warn(
+                "\nWARNING: Note that encoded columns are not guaranteed to reverse transform if they have imputed values.\n"
+                + "If you'd like to use a different type of encoding before imputation, consider using the Pipeline class and call your own encoder before calling this subclass.",
+            )
+            sys.stderr.flush()
+
+            self.ordinal_encoder = EncoderOrdinal(df=df_valid, col_encode=all_cat_cols, unknown_value=np.nan)
+            self.ordinal_encoder.fit()
+            df_to_transf = self.ordinal_encoder.transform(df_valid)
+
+        missing_value_cols = [
+            value
+            for value in list(df_to_transf)
+            if (self.knn_params["missing_values"] in df_to_transf[value] or df_to_transf[value].isna().any())
+        ]
+        missing_value_cols_no_impute = list(set(missing_value_cols) - set(self.col_impute))
+
+        transf_df = pd.DataFrame(self.imputer.transform(df_to_transf), columns=self.valid_cols)
+
+        # attempt inverse_transform encoded imputed columns.
+        try:
+            if self.enable_encoder is True:
+                decoded_transf_df = self.ordinal_encoder.inverse_transform(transf_df)
+                transf_df[cat_cols_missing_value_impute] = decoded_transf_df[cat_cols_missing_value_impute]
+                self.print_message(f"\nImputed categorical columns' encoding was reverse transformed.")
+        except IndexError:
+            self.print_message(
+                f"\nImputed categorical columns' encoding was not reverse transformed."
+                + f"Note that encoded columns are not guaranteed to reverse transform if they have imputed values.\n"
+            )
+            pass
+
+        # reverse cols with missing values not in col_impute to their original values.
+        transf_df[missing_value_cols_no_impute] = df[missing_value_cols_no_impute]
+
+        # inverse_transform cat cols not in col_impute.
+        if self.enable_encoder is True and len(cat_cols_no_missing_value) > 0:
+            decoded_df = self.ordinal_encoder.inverse_transform(df_to_transf)
+            transf_df[cat_cols_no_missing_value] = decoded_df[cat_cols_no_missing_value]
+
+        # re-add categorical cols possibly excluded at fit time with enable_encoder=False.
+        for col in non_valid_cols:
+            transf_df[col] = df[col]
+
+        return transf_df
```

Comment:
A good part of the code here is very similar (or exactly the same) as the iterative imputer. It seems like you could restructure some of the code here, in which you can put some of the code in a private method from the DataImputer class, and then call these methods from the IterativeDataImputer and from the KNN Imputer class.

Also, many  of the comments I did for the iterative imputer applies here as well.

Reply:
good idea!